Issue 95

JonasHoefer opened the pull request:
### Issue 

Closes #95

### Description of the Change

Adds IR to lifted IR translation for types and refactors the Agda back end to use this new intermediate stage.

### Alternate Designs

Parts of the logic (e.g. choosing which translation to use, finding decreasing arguments) must remain in the back end for the moment. Introducing `Declaration`s for the lifted IR and moving dependency analysis to an early stage in the pipeline (see #72) might allow us to move more logic from the back ends to the _IR to lifted IR_ translation.

### Verification Process

The old Agda back end tests were reused to test the new lifted IR functions.

JonasHoefer committed:
Merge branch 'main' into issue-95

JonasHoefer committed:
Merge branch 'main' into issue-95

JonasHoefer committed:
Merge branch 'dev-agda' into issue-95

JonasHoefer committed:
Update test Agda case #95

The old Agda back end annotated all constructor with a Size annotation,
because it was convenient to do so. The new back end places annotations
only if needed. This change does not effect the behavior in Agda.

JonasHoefer committed:
Add IR to lifted IR translation for types #95

JonasHoefer committed:
Refactor Agda back end to use lifted IR #95

JonasHoefer committed:
Refactor lifted IR function translation #95

For reuse in the Coq back end this commit refactors `convertFuncType`.

JonasHoefer committed:
Merge branch 'dev-agda' into issue-95

nbun commented on src/lib/FreeC/Backend/Agda/Converter/Type.hs:
```diff
@@ -1,134 +1,89 @@
-appliesTo (IR.TypeApp _ l _    ) name = l `appliesTo` name
-appliesTo (IR.TypeCon _ conName) name = conName == name
-appliesTo _                      _    = False
+--   If a @Size@ variable is needed, but non is provided an error will be occur.
```
```suggestion
--   If a @Size@ variable is needed, but none is provided, an error occurs.
```

nbun commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,94 @@
+import           FreeC.IR.SrcSpan               ( SrcSpan(NoSrcSpan) )
+import qualified FreeC.LiftedIR.Syntax         as LIR
+
+-- | Functions not defined in terms lambdas (@f x = …@ not @f = \x -> …@) aren't
```
```suggestion
-- | Functions not defined in terms of lambdas (@f x = …@ not @f = \x -> …@) aren't
```

nbun commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,94 @@
+  let startArgs = map convertType $ take (decIndex - 1) args
+      decArg    = markOutermostDecreasing $ convertType $ args !! decIndex
+      endArgs   = map convertType $ drop (decIndex + 1) args
+  in  startArgs ++ (decArg : endArgs)
```
```suggestion
  let convArgs = map convertType args
      (startArgs, decArg:endArgs) = splitAt (decIndex - 1) convArgs
  in  startArgs ++ (markOutermostDecreasing decArg : endArgs)
```

nbun commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,94 @@
+-- Translations                                                              --
+-------------------------------------------------------------------------------
+
+-- | Converts a type from IR to Lifted IR by lifting it into the @Free@ monad.
```
```suggestion
-- | Converts a type from IR to lifted IR by lifting it into the @Free@ monad.
```

nbun commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,94 @@
+convertType :: IR.Type -> LIR.Type
+convertType = LIR.FreeTypeCon NoSrcSpan . convertType'
+
+-- | Lifts a type from IR to Lifted IR by renaming type variables and
```
```suggestion
-- | Lifts a type from IR to lifted IR by renaming type variables and
```

nbun commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,94 @@
+-- | Marks the outermost occurring type constructor as decreasing.
+--
+--   Note: Could be generalized to annotate a constructor based on a position.
+--   At the moment this isn't needed, because our termination checker doesn't
+--   cover cases, where the decreasing argument is just part of anther argument.
+--   e.g. Decreasing in the first element of a pair.
```
```suggestion
--   At the moment this isn't needed because our termination checker doesn't
--   cover cases where the decreasing argument is part of another argument.
--   For example, a decreasing argument in the first element of a pair is not covered.
```

nbun commented on src/lib/FreeC/LiftedIR/Syntax/Type.hs:
```diff
@@ -44,3 +44,23 @@ data Type
+-- Utility                                                                   --
+-------------------------------------------------------------------------------
+
+-- | Applies a type level expression lifted IR expression to a type constructor.
```
Usually you apply a constructor to a type and not the other way around, if that's what is meant here?

nbun commented on src/lib/FreeC/LiftedIR/Syntax/Type.hs:
```diff
@@ -44,3 +44,23 @@ data Type
+func :: Type -> Type -> Type
+func = FuncType NoSrcSpan
+
+-- | Whether the type contains a decreasing argument.
```
```suggestion
-- | Decides whether a type contains a decreasing argument.
```

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Fix typos #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Simplify code #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

JonasHoefer committed:
Rename `convertConArg` #95

JonasHoefer committed:
Merge branch 'issue-95' of github.com:FreeProving/free-compiler into issue-95

JonasHoefer committed:
Reword comment #95

JonasHoefer committed:
Add missing import

JonasHoefer committed:
Improve function names and documentation #95

JonasHoefer committed:
Add missing module documentation #95

nbun commented on src/lib/FreeC/LiftedIR/Syntax/Type.hs:
```diff
@@ -44,3 +44,23 @@ data Type
+-- Utility                                                                   --
+-------------------------------------------------------------------------------
+
+-- | Applies a lifted IR type constructor to type level expression.
```
```suggestion
-- | Applies a lifted IR type constructor to a type level expression.
```

JonasHoefer committed:
Fix typo #95

Co-authored-by: Niels Bunkenburg <30875649+nbun@users.noreply.github.com>

just95 commented on src/test/FreeC/Backend/Agda/Converter/TypeDeclTests.hs:
```diff
@@ -54,7 +54,7 @@ testConvertDataDecls =
         (Recursive ["data List a = Nil | Cons a (List a)"])
         [ "data List (Shape : Set)(Pos : Shape → Set)(a : Set) : {Size} → Set where"
-        ++ "  nil : ∀ {i} → List Shape Pos a {↑ i}"
+        ++ "  nil : List Shape Pos a"
```
In the `free-proving-agda-code` repository we wrote [`nil : ∀ {i} → List S P A {i}`](https://github.com/FreeProving/free-proving-agda-code/blob/8247549708d3a01a78d0e7afbb79f8f57e4a9bd2/src/Free/List.agda#L12).  Have you tested whether `{↑ i}` works as well?

JonasHoefer replied:
Yes, i did some test with the old Ada backend. This new implementation drops the whole annotation if it isn't needed. I tested multiple data types from the `free-proving-agda-code` repository and have not found any problems.
> The old Agda back end annotated all constructor with a Size annotation,
> because it was convenient to do so. The new back end places annotations
> only if needed. This change does not effect the behavior in Agda.

just95 commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,105 @@
+--
+--   > τ' = Free τ*
+convertType :: IR.Type -> LIR.Type
+convertType = LIR.FreeTypeCon NoSrcSpan . convertType'
```
In order to distinguish functions that convert IR to lifted IR from the functions that convert lifted IR to the target language, I think that we should give them different names. Especially since the `convert` prefix suggests that a function is in the `Converter` moand, I would use that prefix in this case. I'd recommend the following naming scheme.

 - `lift*` for the translation from IR to lifted IR.
 - `convertLifted*` for the translation from lifted IR to the intermediate language.
 - `convert*` for the translation from the IR to the target language (i.e., `convertLifted* . lift*`).

This will also make the qualified import of the IR to lifted IR conversion functions unnecesarry.

just95 commented on src/lib/FreeC/LiftedIR/Syntax/Type.hs:
```diff
@@ -44,3 +44,23 @@ data Type
+-- Utility                                                                   --
+-------------------------------------------------------------------------------
+
+-- | Applies a lifted IR type constructor to a type level expression.
+typeApp :: Type -> Type -> Type
+typeApp (TypeCon srcSpan name ts dec) t = TypeCon srcSpan name (t : ts) dec
+typeApp _ _ = error "Cannot apply type to non type constructor!"
```
The problem I've got with this function is that `typeApp (D t₁ … tₙ) t₀ = D t₀ t₁ … tₙ`. That's why `LIR.convertType'` reverses the argument list and why you have to apply `reverse` in `Agda.convertType`. The reversing of arguments is a potential source of errors and not documented anywhere. I think that `LIR.convertType'` should be implemented with an accumulator internally such that the order of arguments is preserved.

just95 commented on src/lib/FreeC/LiftedIR/Syntax/Type.hs:
```diff
@@ -44,3 +44,23 @@ data Type
+typeApp (TypeCon srcSpan name ts dec) t = TypeCon srcSpan name (t : ts) dec
+typeApp _ _ = error "Cannot apply type to non type constructor!"
+
+-- | Smart constructor for lifted IR functions.
+func :: Type -> Type -> Type
+func = FuncType NoSrcSpan
```
I don't like smart constructors that just drop the `SrcSpan` argument. They encourage the user to drop location information. In the IR there is for example the following smart constructor which does exactly what you use `LIR.func` for, but does force the user to think about relevant location information (or actively decide that there is no `SrcSpan` than can be inserted).

```haskell
-- | Creates a function type with the given argument and return types.
funcType :: SrcSpan -> [Type] -> Type -> Type
funcType srcSpan = flip (foldr (FuncType srcSpan))
```

Also note that smart constructor names should match the constructor names.

Daniel-Teut commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,105 @@
+convertType :: IR.Type -> LIR.Type
+convertType = LIR.FreeTypeCon NoSrcSpan . convertType'
+
+-- | Lifts a type from IR to lifted IR by renaming type variables and
```
As far as I can see `convertType'` does not actually rename any type variables or constructors.

JonasHoefer committed:
Improve LIR function type smart constructor #95

JonasHoefer committed:
`LIR.TypeCon` ow uses a snoc list #95

The old version of `LIR.TypeCon` uses a normal list to store arguments. To apply
the arguments in the right order either a `foldr` with a flipped function or
`foldl` with a reversed list has to be used. Furthermore the arguments in the
list are stored in the conceptually wrong order. This suggests that the list
should actually be a snoc list, which internally already has the right structure
and therefore exactly the needed recursion scheme.

    :          $
   / \        / \
  :   x  =>  $   x
 / \        / \
... y      ... y

JonasHoefer committed:
Rename conversion/lifting functions #95

just95 commented on src/lib/FreeC/LiftedIR/Converter/Type.hs:
```diff
@@ -0,0 +1,105 @@
+--   > (τ₁ -> τ₂)* = τ₁' -> τ₂'
+--   > C* = Ĉ Shape Position
+--   > α* = α̂
+liftType' :: IR.Type -> LIR.Type
+liftType' (IR.TypeCon srcSpan name) = LIR.TypeCon srcSpan name Nil False
+liftType' (IR.TypeVar srcSpan name) = LIR.TypeVar srcSpan name
+liftType' (IR.TypeApp _ l r       ) = liftType' l `LIR.typeApp` liftType' r
+liftType' (IR.FuncType srcSpan l r) =
+  LIR.FuncType srcSpan (liftType l) (liftType r)
```
I don't think that using a `SnocList` for `typeConArgs` really solves the underlying problem that I've pointed out in my [previous comment](https://github.com/FreeProving/free-compiler/pull/103#discussion_r448538273). On the upside, `SnocList` makes explicit that the list is in reverse order. On the downside, the list is effectively still stored in the wrong order.

Consider the following grammar for the lifted IR, for example.

```haskell
Type ::= ConIdent { Type } | … 
```

Parsers and pretty printers for this grammar still have to reverse the arguments `{ Type }`. That's odd but may be justified if it simplified operations on the lifted IR. 

From the point of view of `convertLiftedType`, it does not matter whether the arguments are stored as a list or `SnocList`. Only `Foldable` and `Traversable` instances are needed. However, since those instances are not tail recursive in case of `SnocList`, performance would be better when `typeConArgs` were stored as a list. 

The only code that really benifits from a `SnocList` is `liftType'`. However, `liftType'` can easily be written in terms of a helper function that accumulates the arguments.

```suggestion
liftType' :: IR.Type -> LIR.Type
liftType' = flip liftTypeApp' []

-- | Like 'liftType'' but accumulates the type arguments in a second parameter.
--
--   > (C τ₁ … τₙ)* = C τ₁* … τₙ*
--   > (τ₁ -> τ₂)* = τ₁' -> τ₂'
--   > α* = α
liftTypeApp' :: IR.Type -> [IR.Type] -> LIR.Type
liftTypeApp' (IR.TypeCon srcSpan name) ts =
  LIR.TypeCon srcSpan name (map liftType' ts) False
liftTypeApp' (IR.TypeVar srcSpan name) [] = LIR.TypeVar srcSpan name
liftTypeApp' (IR.TypeApp _ l r       ) ts = liftTypeApp' l (r : ts)
liftTypeApp' (IR.FuncType srcSpan l r) [] =
  LIR.FuncType srcSpan (liftType l) (liftType r)
liftTypeApp' _ (_ : _) = error "liftTypeApp': Only type constructors can be applied!"
```

Compared to the `SnocList` approach, much less code has to be written (`typeApp` can even be deleted), the argument list does not have to be reversed in any of our use cases and the code in `convertLiftedType` gets slightly more performant.

JonasHoefer replied:
My problem with this approach is approach is that it also doesn't solve the problem and just moves it. Storing the type parameters in the 'correct' order forces us to use `foldl` to build the AST in the backend, therefore just reversing the list again. Even worse, the IR already has the right structure! This approach just reverses the structure twice while passing it through the pipeline.
```
     ⋰                                        ⋰
    @                         :              $  
   / \                       / \            / \ 
  @   τ₂   =>  IR.TypeApp C τ₁  :    =>    $   τ₂  
 / \                           / \        / \   
C   τ₁                        τ₂  …      C   τ₁ 
```
Storing the list in 'reverse' order (or as a snoc list to just visually flip it) doesn't change the structure, allowing us to use the right recursion scheme to build the AST in the back end (`foldr` for `[LIR.Type]` or `foldl` for `SnocList LIR.Type`), building the AST without reversing the list. 

Since parsing and pretty printing are left to right operations we have to deal with the problem of reversing the list at some point. I think we should just solve the problem as early as possible to avoid influencing the main flow of data (i.e. reversing all type applications twice while compiling).

Dealing with left recursion while parsing shouldn't be a problem. We can either parse the structure right recursive and reverse it later (`reverse` or `foldl`) or use [`chainl`](https://hackage.haskell.org/package/parsec-3.1.14.0/docs/src/Text.Parsec.Combinator.html#chainl), which avoids unnecessary backtracking by parsing bottom up. Note that dealing with the problem while parsing avoids the `foldl` in the backend.

While pretty printing we of course have to deal with the problem again. One way of reducing the thunk buildup of `foldl` is to use [`foldl'`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:foldl-39-) instead.

---

From a performance point of view (which shouldn't really matter, because our list has like 5 items :smile:) it's just a trade of between reversing during IO or between pipeline stages. 

From an elegance/readability point of view I'm not quite sure which approach is better. I think I lean towards the simpler algorithms working on the expected structure. As a side note: I already started working on the conversion of expressions. The Coq backend uses the same technique of accumulating arguments which I reused, because in this case it simplifies the algorithm.

just95 replied:
> performance [...] shouldn't really matter, because our list has like 5 items

I agree. Performance is not really a priority for us. But why do we then care about the implicit reversals performed by `foldl` in the back end? The point I wanted to make with my performance argument is that the `SnocList` approach complicates the code but doesn't even improve performance.

> Since parsing and pretty printing are left to right operations we have to deal with the problem of reversing the list at some point.

I've included the parsing and pretty printing example to show that even though `liftType'` can be defined slightly more compactly with `SnocList`s, other code will become more complex. For parsing a list of arguments, we can simply use `Parsec.many` and `hsep . map pretty` for pretty printing. I guess that much more code is needed to parse or pretty print a `SnocList`.

> The Coq backend uses the same technique of accumulating arguments

That's only partially true. Yes, I don't use an accumulator argument in [`convertType'`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Converter/Type.hs#L44). However, unlike `Agda.typeApp`, [`Coq.app`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Syntax.hs#L92) adds the type argument to the back of the list and not to the front. If I recommended to use `(++)` instead of `(:)` in `Agda.typeApp`, I'd most likely have started a discussion about performance. That's one reason why I've suggested to use an accumulator argument. The other reason is that [`convertExpr'`](https://github.com/FreeProving/free-compiler/blob/35bbb90b0c5d776489dae00539b2b55acbfb0ae1/src/lib/FreeC/Backend/Coq/Converter/Expr.hs#L41) requires an accumulator argument. Therefore, adding an accumulator argument to `liftType'` improves consistency.

just95 commented on src/lib/FreeC/Backend/Agda/Converter/FuncDecl.hs:
```diff
@@ -13,15 +13,14 @@ import           FreeC.Backend.Agda.Converter.Arg
 import           FreeC.Environment.LookupOrFail
 import           FreeC.IR.DependencyGraph
 import qualified FreeC.IR.Syntax               as IR
+import qualified FreeC.LiftedIR.Converter.Type as LIR
```
After renaming the lifting functions, it is not necessary anymore to qualify imports of `FreeC.LiftedIR.Converter.*`.

```suggestion
import           FreeC.LiftedIR.Converter.Type
```

JonasHoefer committed:
Remove qualified `LIR` imports #95

JonasHoefer committed:
Remove Snoc List #95

We decided that using a normal list, storing the elements in the written order,
is easier than introducing a new data structure, because standard list integrate
better with other libraries (e.g. pretty printer, parser).

JonasHoefer committed:
Format code #95
